File: src/main/java/com/sgx/icms/service/ClaimService.java
package com.sgx.icms.service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.sgx.icms.dao.ClaimDao;
import com.sgx.icms.dao.DocumentDao;
import com.sgx.icms.dao.PolicyDao;
import com.sgx.icms.dao.PolicyholderDao;
import com.sgx.icms.db.Db;
import com.sgx.icms.domain.AuditLog;
import com.sgx.icms.domain.Claim;
import com.sgx.icms.domain.ClaimDocument;
import com.sgx.icms.domain.ClaimStatus;
import com.sgx.icms.domain.DocumentRequirement;
import com.sgx.icms.domain.Policy;
import com.sgx.icms.domain.Policyholder;
import com.sgx.icms.domain.Roles;
import com.sgx.icms.web.support.Paged;
import com.sgx.icms.web.support.SessionUser;
import com.sgx.icms.web.support.TimelineStage;

/**
 * Claim use-cases for the customer portal: resolve the customer's policyholder,
 * list/read their claims (ownership-enforced), create drafts/submissions (with
 * required-document seeding and agent notification), and withdraw.
 *
 * <p>Status transitions are guarded; multi-table writes run in one transaction.
 */
public class ClaimService {

    private static final Logger LOG = LoggerFactory.getLogger(ClaimService.class);

    private static final List<String> OPEN_STATUSES = Arrays.asList(
            ClaimStatus.DRAFT, ClaimStatus.SUBMITTED, ClaimStatus.UNDER_REVIEW,
            ClaimStatus.SURVEY_SCHEDULED, ClaimStatus.UNDER_ASSESSMENT, ClaimStatus.PENDING_APPROVAL,
            ClaimStatus.APPROVED, ClaimStatus.SETTLEMENT_PROCESSING, ClaimStatus.ON_HOLD);
    private static final List<String> SETTLED_STATUSES = Arrays.asList(
            ClaimStatus.SETTLED, ClaimStatus.CLOSED);

    private final ClaimDao claimDao = new ClaimDao();
    private final PolicyDao policyDao = new PolicyDao();
    private final PolicyholderDao policyholderDao = new PolicyholderDao();
    private final DocumentDao documentDao = new DocumentDao();
    private final AuditService audit = new AuditService();
    private final NotificationService notifications = new NotificationService();
    private final CommunicationService communications = new CommunicationService();

    /* ----------------------------- reads ----------------------------- */

    public Policyholder resolveCustomer(String email) {
        return Db.withConnection(conn -> policyholderDao.findByEmail(conn, email));
    }

    public List<Policy> policiesForCustomer(long policyholderId) {
        return Db.withConnection(conn -> policyDao.findByPolicyholder(conn, policyholderId));
    }

    public Paged<Claim> listForCustomer(long policyholderId, int page, int size) {
        int offset = (page - 1) * size;
        return Db.withConnection(conn -> {
            long total = claimDao.countByPolicyholder(conn, policyholderId);
            List<Claim> items = claimDao.findByPolicyholder(conn, policyholderId, size, offset);
            return new Paged<>(items, page, size, total);
        });
    }

    /** [total, open, settled] counts for the customer dashboard. */
    public long[] customerCounts(long policyholderId) {
        return Db.withConnection(conn -> new long[] {
                claimDao.countByPolicyholder(conn, policyholderId),
                claimDao.countByPolicyholderInStatuses(conn, policyholderId, OPEN_STATUSES),
                claimDao.countByPolicyholderInStatuses(conn, policyholderId, SETTLED_STATUSES)
        });
    }

    /** Returns the claim only if it belongs to this policyholder (authorisation). */
    public Claim getOwnedClaim(long policyholderId, long claimId) {
        Claim c = Db.withConnection(conn -> claimDao.findById(conn, claimId));
        if (c == null || c.getPolicyholderId() != policyholderId) {
            return null;
        }
        return c;
    }

    /* ----------------------------- writes ----------------------------- */

    /**
     * Creates a claim (draft or submitted) for the given policyholder. Seeds the
     * required-document checklist and, on submit, notifies agents — all atomically.
     *
     * @return the new claim id
     * @throws IllegalArgumentException if the chosen policy is not the customer's
     */
    public long createClaim(SessionUser actor, Policyholder ph, Claim draft, boolean submit, String ip) {
        final int year = LocalDate.now().getYear();
        return Db.inTransaction(conn -> {
            Policy policy = policyDao.findById(conn, draft.getPolicyId());
            if (policy == null || policy.getPolicyholderId() != ph.getId()) {
                throw new IllegalArgumentException("Please select one of your own policies.");
            }

            draft.setPolicyholderId(ph.getId());
            if (draft.getClaimantName() == null || draft.getClaimantName().trim().isEmpty()) {
                draft.setClaimantName(ph.getFullName());
            }
            draft.setClaimType(policy.getType()); // claim type follows the policy type
            draft.setStatus(submit ? ClaimStatus.SUBMITTED : ClaimStatus.DRAFT);
            draft.setRiskLevel("LOW");
            draft.setFraudScore(0);
            draft.setClaimNo(claimDao.nextClaimNo(conn, year));

            long claimId = claimDao.insert(conn, draft);
            seedRequiredDocuments(conn, claimId, draft.getClaimType(), draft.getClaimSubtype());

            String action = submit ? "CLAIM_SUBMITTED" : "CLAIM_DRAFT_SAVED";
            if (submit) {
                notifications.notifyRole(conn, Roles.AGENT, "ACTION",
                        "New claim " + draft.getClaimNo() + " submitted and awaiting acknowledgement.");
                communications.system(conn, claimId, "ICMS",
                        "Your claim " + draft.getClaimNo() + " has been received and is awaiting review.");
            }
            writeAudit(conn, actor, action, draft.getClaimNo(), ip);

            LOG.info("Customer {} {} claim {} (id={})", actor.getUsername(), action, draft.getClaimNo(), claimId);
            return claimId;
        });
    }

    /** Customer withdraws their own claim if its status permits. */
    public boolean withdraw(SessionUser actor, long policyholderId, long claimId, String ip) {
        return Db.inTransaction(conn -> {
            Claim c = claimDao.findById(conn, claimId);
            if (c == null || c.getPolicyholderId() != policyholderId) {
                return false; // not owner
            }
            if (!ClaimStatus.isWithdrawable(c.getStatus())) {
                throw new IllegalStateException("This claim can no longer be withdrawn.");
            }
            claimDao.updateStatus(conn, claimId, ClaimStatus.WITHDRAWN);
            notifications.notifyRole(conn, Roles.AGENT, "INFO",
                    "Claim " + c.getClaimNo() + " was withdrawn by the customer.");
            writeAudit(conn, actor, "CLAIM_WITHDRAWN", c.getClaimNo(), ip);
            return true;
        });
    }

    /* ----------------------------- helpers ----------------------------- */

    private void seedRequiredDocuments(java.sql.Connection conn, long claimId, String type, String subtype) {
        List<DocumentRequirement> reqs = documentDao.findRequirements(conn, type, subtype);
        for (DocumentRequirement r : reqs) {
            ClaimDocument d = new ClaimDocument();
            d.setClaimId(claimId);
            d.setDocType(r.getDocType());
            d.setUploadStatus(r.isRequired() ? "PENDING" : "CONDITIONAL");
            d.setVerificationStatus("PENDING");
            documentDao.insert(conn, d);
        }
    }

    private void writeAudit(java.sql.Connection conn, SessionUser actor, String action, String claimNo, String ip) {
        AuditLog a = new AuditLog();
        a.setUserId(actor.getId());
        a.setUsername(actor.getUsername());
        a.setRole(actor.getRole());
        a.setAction(action);
        a.setEntity(claimNo);
        a.setIpAddress(ip);
        a.setResult(AuditLog.RESULT_SUCCESS);
        audit.record(conn, a);
    }

    /** Builds the status timeline view model for a claim. */
    public List<TimelineStage> timeline(Claim c) {
        List<TimelineStage> stages = new ArrayList<>();
        String status = c.getStatus();

        if (ClaimStatus.DRAFT.equals(status)) {
            stages.add(new TimelineStage("Draft", TimelineStage.CURRENT));
            for (String s : ClaimStatus.LIFECYCLE) {
                stages.add(new TimelineStage(ClaimStatus.label(s), TimelineStage.PENDING));
            }
            return stages;
        }

        int curIdx;
        if (ClaimStatus.ON_HOLD.equals(status)) {
            curIdx = ClaimStatus.LIFECYCLE.indexOf(ClaimStatus.PENDING_APPROVAL);
        } else if (ClaimStatus.REJECTED.equals(status) || ClaimStatus.WITHDRAWN.equals(status)) {
            curIdx = -1; // off-path terminal; the status pill conveys the real state
        } else {
            curIdx = ClaimStatus.lifecycleIndex(status);
        }

        for (int i = 0; i < ClaimStatus.LIFECYCLE.size(); i++) {
            String state;
            if (curIdx < 0) {
                state = TimelineStage.PENDING;
            } else if (i < curIdx) {
                state = TimelineStage.DONE;
            } else if (i == curIdx) {
                state = TimelineStage.CURRENT;
            } else {
                state = TimelineStage.PENDING;
            }
            stages.add(new TimelineStage(ClaimStatus.label(ClaimStatus.LIFECYCLE.get(i)), state));
        }
        return stages;
    }
}

File: backend/src/services/claimService.js
'use strict';

const { knex } = require('../db/knex');
const { withTransaction } = require('../db/tx');
const claimRepo = require('../repositories/claimRepo');
const policyRepo = require('../repositories/policyRepo');
const policyholderRepo = require('../repositories/policyholderRepo');
const documentService = require('./documentService');
const communicationService = require('./communicationService');
const notificationService = require('./notificationService');
const audit = require('./auditService');
const status = require('../domain/claimStatus');
const { paged } = require('../utils/paging');
const { ValidationError, ConflictError } = require('../utils/errors');
const logger = require('../config/logger');

const AGENT = 'AGENT';

const OPEN_STATUSES = [
  status.DRAFT,
  status.SUBMITTED,
  status.UNDER_REVIEW,
  status.SURVEY_SCHEDULED,
  status.UNDER_ASSESSMENT,
  status.PENDING_APPROVAL,
  status.APPROVED,
  status.SETTLEMENT_PROCESSING,
  status.ON_HOLD,
];
const SETTLED_STATUSES = [status.SETTLED, status.CLOSED];

/** Augment a raw claim row with the view-model fields the JSP getters provided. */
function withStatusView(c) {
  if (!c) return c;
  return { ...c, statusLabel: status.label(c.status), statusPill: status.pill(c.status) };
}

/** Customer-safe projection — omits internal-only fields (notes, fraud, risk, ids). */
function toCustomerClaim(c) {
  if (!c) return null;
  const v = withStatusView(c);
  const {
    internalNotes,
    fraudScore,
    riskLevel,
    agentId,
    surveyorId,
    surveyorName,
    slaDueDate,
    ...safe
  } = v;
  return { ...safe, withdrawable: status.isWithdrawable(c.status) };
}

/* ----------------------------- reads ----------------------------- */

async function resolveCustomer(email) {
  return policyholderRepo.findByEmail(knex, email);
}

async function policiesForCustomer(policyholderId) {
  const rows = await policyRepo.findByPolicyholder(knex, policyholderId);
  return rows.map((p) => ({ ...p, displayLabel: `${p.policyNo} — ${p.type}` }));
}

async function listForCustomer(policyholderId, page, size, offset) {
  const [total, rows] = await Promise.all([
    claimRepo.countByPolicyholder(knex, policyholderId),
    claimRepo.findByPolicyholder(knex, policyholderId, size, offset),
  ]);
  return paged(rows.map(toCustomerClaim), page, size, total);
}

/** [total, open, settled] for the customer dashboard. */
async function customerCounts(policyholderId) {
  const [total, open, settled] = await Promise.all([
    claimRepo.countByPolicyholder(knex, policyholderId),
    claimRepo.countByPolicyholderInStatuses(knex, policyholderId, OPEN_STATUSES),
    claimRepo.countByPolicyholderInStatuses(knex, policyholderId, SETTLED_STATUSES),
  ]);
  return { total, open, settled };
}

/** Returns the claim only if it belongs to this policyholder (authorisation). */
async function getOwnedClaim(policyholderId, claimId) {
  const c = await claimRepo.findById(knex, claimId);
  if (!c || Number(c.policyholderId) !== Number(policyholderId)) return null;
  return c;
}

/** Full customer claim detail bundle: claim + timeline + documents + messages. */
async function customerClaimBundle(policyholderId, claimId) {
  const raw = await getOwnedClaim(policyholderId, claimId);
  if (!raw) return null;
  const [documents, messages] = await Promise.all([
    documentService.forClaim(claimId),
    communicationService.forClaim(claimId),
  ]);
  return {
    claim: toCustomerClaim(raw),
    timeline: status.timeline(raw.status),
    documents,
    messages,
  };
}

/* ----------------------------- writes ----------------------------- */

/**
 * Creates a claim (draft or submitted). Seeds required docs and, on submit,
 * notifies agents + posts a system message — all atomically. Port of
 * ClaimService.createClaim.
 * @returns {Promise<{id:number, claimNo:string}>}
 */
async function createClaim(actor, ph, draft, submit, ip) {
  const year = new Date().getFullYear();
  return withTransaction(async (trx) => {
    const policy = await policyRepo.findById(trx, draft.policyId);
    if (!policy || Number(policy.policyholderId) !== Number(ph.id)) {
      throw new ValidationError('Please select one of your own policies.');
    }

    draft.policyholderId = ph.id;
    if (!draft.claimantName || !draft.claimantName.trim()) draft.claimantName = ph.fullName;
    draft.claimType = policy.type; // claim type follows the policy type
    draft.status = submit ? status.SUBMITTED : status.DRAFT;
    draft.riskLevel = 'LOW';
    draft.fraudScore = 0;
    draft.claimNo = await claimRepo.nextClaimNo(trx, year);

    const claimId = await claimRepo.insert(trx, draft);
    await documentService.seedRequiredDocuments(trx, claimId, draft.claimType, draft.claimSubtype);

    const action = submit ? 'CLAIM_SUBMITTED' : 'CLAIM_DRAFT_SAVED';
    if (submit) {
      await notificationService.notifyRole(
        trx,
        AGENT,
        'ACTION',
        `New claim ${draft.claimNo} submitted and awaiting acknowledgement.`
      );
      await communicationService.system(
        trx,
        claimId,
        'ICMS',
        `Your claim ${draft.claimNo} has been received and is awaiting review.`
      );
    }
    await audit.record(trx, {
      userId: Number(actor.id),
      username: actor.username,
      role: actor.role,
      action,
      entity: draft.claimNo,
      ipAddress: ip,
      result: 'SUCCESS',
    });

    logger.info({ user: actor.username, action, claimNo: draft.claimNo, claimId }, 'Claim created');
    return { id: claimId, claimNo: draft.claimNo };
  });
}

/** Customer withdraws their own claim if status permits. Port of ClaimService.withdraw. */
async function withdraw(actor, policyholderId, claimId, ip) {
  return withTransaction(async (trx) => {
    const c = await claimRepo.findById(trx, claimId);
    if (!c || Number(c.policyholderId) !== Number(policyholderId)) return false; // not owner
    if (!status.isWithdrawable(c.status)) {
      throw new ConflictError('This claim can no longer be withdrawn.');
    }
    await claimRepo.updateStatus(trx, claimId, status.WITHDRAWN);
    await notificationService.notifyRole(
      trx,
      AGENT,
      'INFO',
      `Claim ${c.claimNo} was withdrawn by the customer.`
    );
    await audit.record(trx, {
      userId: Number(actor.id),
      username: actor.username,
      role: actor.role,
      action: 'CLAIM_WITHDRAWN',
      entity: c.claimNo,
      ipAddress: ip,
      result: 'SUCCESS',
    });
    return true;
  });
}

module.exports = {
  resolveCustomer,
  policiesForCustomer,
  listForCustomer,
  customerCounts,
  getOwnedClaim,
  customerClaimBundle,
  createClaim,
  withdraw,
  timeline: status.timeline,
  withStatusView,
  toCustomerClaim,
};
